src/theory/strings/solver_state.h

```diff
@@ -0,0 +1,184 @@
+   * the theory of strings.
+   */
+  void setConflict();
+  /** Are we currently in conflict? */
+  bool isInConflict() const;
+  /** set pending conflict
+   *
+   * If conf is non-null, this is called when conf is a conjunction of literals
+   * that hold in the current context that are unsatisfiable. It is set as the
+   * "pending conflict" to be processed as a conflict lemma on the output
+   * channel of this class. It is not sent out immediately since it may require
+   * explanation from the equality engine, and may be called at any time, e.g.
+   * during a merge operation, when the equality engine is not in a state to
+   * provide explanations.
+   */
+  void setPendingConflictWhen(Node conf);
+  /** get the pending conflict, or null if none exist */
+  Node getPendingConflict() const;
+  //------------------------------------------ end conflicts
+  /** get length with explanation
+   *
+   * If possible, this returns an arithmetic term that exists in the current
+   * context that is equal to the length of te, or otherwise returns the
+   * length of t. It adds to exp literals that hold in the current context that
+   * explain why that term is equal to the length of t. For example, if
+   * we have assertions:
+   *   len( x ) = 5 ^ z = x ^ x = y,
+   * then getLengthExp( z, exp, y ) returns len( x ) and adds { z = x } to
+   * exp. On the other hand, getLengthExp( z, exp, x ) returns len( x ) and
+   * adds nothing to exp.
+   */
+  Node getLengthExp(Node t, std::vector<Node>& exp, Node te);
+  /** shorthand for getLengthExp(t, exp, t) */
+  Node getLength(Node t, std::vector<Node>& exp);
+  /**
+   * Get the above information for equivalence class eqc. If doMake is true,
+   * we construct a new information class if one does not exist. The term eqc
+   * should currently be a representative of the equality engine of this class.
+   */
+  EqcInfo* getOrMakeEqcInfo(Node eqc, bool doMake = true);
+
+  /** add endpoints to eqc info
+   *
+   * This method is called when term t is the explanation for why equivalence
+   * class eqc may have a constant endpoint due to a concatentation term concat.
+   * For example, we may call this method on:
+   *   t := (str.++ x y), concat := (str.++ x y), eqc
+   * for some eqc that is currently equal to t. Another example is:
+   *   t := (str.in.re z (re.++ r s)), concat := (re.++ r s), eqc
+   * for some eqc that is currently equal to z.
+   */
+  void addEndpointsToEqcInfo(Node t, Node concat, Node eqc);
+
+ private:
+  /** Pointer to the SAT context object used by the theory of strings. */
+  context::Context* d_context;
+  /** Reference to equality engine of the theory of strings. */
+  eq::EqualityEngine& d_ee;
+  /** Are we in conflict? */
+  context::CDO<bool> d_conflict;
+  /** The pending conflict if one exists */
+  context::CDO<Node> d_pendingConflict;
+  /** Map from representatives to their equivalence class information */
+  std::map<Node, EqcInfo*> d_eqc_info;
```

Comment:
In a future PR, we should probably either make this a `std::map<Node, EqcInfo*>` or a `std::map<Node, std::unique_ptr<EqcInfo>>`.